dev_docs/tutorials/testing_plugins.mdx

```diff
@@ -0,0 +1,1343 @@
+---
+id: kibDevTutorialTestingPlugins
+slug: /kibana-dev-docs/tutorial/testing-plugins
+title: Testing Kibana Plugins
+summary: Learn how to test different aspects of Kibana plugins
+date: 2021-07-05
+tags: ['kibana', 'onboarding', 'dev', 'architecture', 'testing']
+---
+
+This document outlines best practices and patterns for testing Kibana Plugins.
+
+## Strategy
+
+In general, we recommend three tiers of tests:
+- Unit tests: small, fast, exhaustive, make heavy use of mocks for external dependencies 
+- Integration tests: higher-level tests that verify interactions between systems (eg. HTTP APIs, Elasticsearch API calls, calling other plugin contracts).
+- End-to-end tests (e2e): tests that verify user-facing behavior through the browser
+
+These tiers should roughly follow the traditional ["testing pyramid"](https://martinfowler.com/articles/practical-test-pyramid.html), where there are more exhaustive testing at the unit level, fewer at the integration level, and very few at the functional level. 
+
+## New concerns in the Kibana Platform
+
+The Kibana Platform introduces new concepts that legacy plugins did not have concern themselves with. Namely:
+- **Lifecycles**: plugins now have explicit lifecycle methods that must interop with Core APIs and other plugins.
```

Comment:
Should be moved at the bottom or removed. With the migration to the KP finished, it's not that important anymore.